File: connectors/java/src/integrationTest/java/com/playhouse/connector/core/C02_TcpConnectionTests.java
package com.playhouse.connector.core;

import com.playhouse.connector.support.BaseIntegrationTest;
import com.playhouse.connector.support.CreateStageResponse;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * C-02: TCP 연결 테스트
 * <p>
 * Connector가 테스트 서버의 TCP 포트(28001)에 성공적으로 연결할 수 있는지 검증합니다.
 * </p>
 */
@DisplayName("C-02: TCP Connection Tests")
public class C02_TcpConnectionTests extends BaseIntegrationTest {

    @Test
    @DisplayName("C-02-01: Stage 생성 후 TCP 연결이 성공한다")
    public void connect_afterStageCreation_succeeds() throws Exception {
        // Given: Stage가 생성되어 있음
        stageInfo = testServer.createTestStage();

        // When: TCP 연결 시도
        connector.setStageId(stageInfo.getStageId());
        CompletableFuture<Void> connectFuture = connector.connectAsync(host, tcpPort);
        connectFuture.get(5, TimeUnit.SECONDS);

        // Then: 연결이 성공해야 함
        assertThat(connector.isConnected()).isTrue();
        assertThat(connector.getStageId()).isEqualTo(stageInfo.getStageId());
    }

    @Test
    @DisplayName("C-02-02: 연결 후 IsConnected는 true를 반환한다")
    public void isConnected_afterConnection_returnsTrue() throws Exception {
        // Given: 연결되지 않은 상태
        assertThat(connector.isConnected()).isFalse();

        // When: 연결 성공
        createStageAndConnect();

        // Then: IsConnected가 true를 반환해야 함
        assertThat(connector.isConnected()).isTrue();
    }

    @Test
    @DisplayName("C-02-03: 연결 전 IsAuthenticated는 false를 반환한다")
    public void isAuthenticated_beforeAuthentication_returnsFalse() throws Exception {
        // Given & When: 연결만 성공한 상태 (인증 전)
        createStageAndConnect();

        // Then: 인증 전이므로 IsAuthenticated는 false여야 함
        assertThat(connector.isAuthenticated()).isFalse();
    }

    @Test
    @DisplayName("C-02-04: OnConnect 이벤트가 성공 결과로 발생한다")
    public void onConnect_event_triggersWithSuccess() throws Exception {
        // Given: Stage 생성
        stageInfo = testServer.createTestStage();

        AtomicBoolean connectResult = new AtomicBoolean(false);
        AtomicBoolean eventTriggered = new AtomicBoolean(false);

        connector.setOnConnect(() -> {
            connectResult.set(true);
            eventTriggered.set(true);
        });

        // When: 연결 시도
        connector.setStageId(stageInfo.getStageId());
        connector.connectAsync(host, tcpPort);

        // OnConnect 이벤트 대기 (MainThreadAction 호출하면서 최대 5초)
        boolean completed = waitForCondition(eventTriggered::get, 5000);

        // Then: 이벤트가 발생하고 성공 결과를 전달해야 함
        assertThat(completed).isTrue();
        assertThat(eventTriggered.get()).isTrue();
        assertThat(connectResult.get()).isTrue();
    }

    @Test
    @DisplayName("C-02-05: 잘못된 Stage ID로 연결해도 TCP 연결은 성공한다")
    public void connect_withInvalidStageId_tcpConnectionSucceeds() throws Exception {
        // Given: 존재하지 않는 Stage ID
        long invalidStageId = 999999999L;

        // When: 잘못된 Stage ID로 연결 시도
        connector.setStageId(invalidStageId);
        CompletableFuture<Void> connectFuture = connector.connectAsync(host, tcpPort);
        connectFuture.get(5, TimeUnit.SECONDS);

        // Then: TCP 연결 자체는 성공함 (Stage ID 검증은 서버 레벨에서 나중에 이루어짐)
        assertThat(connector.isConnected()).isTrue();
        assertThat(connector.isAuthenticated()).isFalse();
    }

    @Test
    @DisplayName("C-02-06: 동일한 Connector로 재연결할 수 있다")
    public void connect_multipleTimes_succeeds() throws Exception {
        // Given: 첫 번째 연결
        createStageAndConnect();
        assertThat(connector.isConnected()).isTrue();

        // When: 연결 해제 후 재연결
        connector.disconnect();
        Thread.sleep(500); // 연결 해제 대기

        CreateStageResponse newStageInfo = testServer.createTestStage();
        connector.setStageId(newStageInfo.getStageId());
        CompletableFuture<Void> reconnectFuture = connector.connectAsync(host, tcpPort);
        reconnectFuture.get(5, TimeUnit.SECONDS);

        // Then: 재연결이 성공해야 함
        assertThat(connector.isConnected()).isTrue();
        assertThat(connector.getStageId()).isEqualTo(newStageInfo.getStageId());
    }
}

File: connectors/csharp/tests/PlayHouse.Connector.IntegrationTests/CoreTests/C02_TcpConnectionTests.cs
using FluentAssertions;
using Xunit;

namespace PlayHouse.Connector.IntegrationTests.CoreTests;

/// <summary>
/// C-02: TCP 연결 테스트
/// </summary>
/// <remarks>
/// Connector가 테스트 서버의 TCP 포트(34001)에 성공적으로 연결할 수 있는지 검증합니다.
/// </remarks>
public class C02_TcpConnectionTests : BaseIntegrationTest
{
    public C02_TcpConnectionTests(TestServerFixture testServer) : base(testServer)
    {
    }

    [Fact(DisplayName = "C-02-01: Stage 생성 후 TCP 연결이 성공한다")]
    public async Task Connect_AfterStageCreation_Succeeds()
    {
        // Given: Stage가 생성되어 있음
        StageInfo = await TestServer.CreateTestStageAsync();

        // When: TCP 연결 시도
        var connected = await Connector!.ConnectAsync(
            TestServer.Host,
            TestServer.TcpPort,
            StageInfo.StageId,
            StageInfo.StageType
        );

        // Then: 연결이 성공해야 함
        connected.Should().BeTrue("TCP 연결이 성공해야 함");
        Connector.IsConnected().Should().BeTrue("연결 상태가 true여야 함");
        Connector.StageId.Should().Be(StageInfo.StageId, "연결된 Stage ID가 일치해야 함");
        Connector.StageType.Should().Be(StageInfo.StageType, "연결된 Stage 타입이 일치해야 함");
    }

    [Fact(DisplayName = "C-02-02: 연결 후 IsConnected는 true를 반환한다")]
    public async Task IsConnected_AfterConnection_ReturnsTrue()
    {
        // Given: 연결되지 않은 상태
        Connector!.IsConnected().Should().BeFalse("초기 상태는 연결되지 않음");

        // When: 연결 성공
        await CreateStageAndConnectAsync();

        // Then: IsConnected가 true를 반환해야 함
        Connector.IsConnected().Should().BeTrue("연결 후 IsConnected는 true여야 함");
    }

    [Fact(DisplayName = "C-02-03: 연결 전 IsAuthenticated는 false를 반환한다")]
    public async Task IsAuthenticated_BeforeAuthentication_ReturnsFalse()
    {
        // Given & When: 연결만 성공한 상태 (인증 전)
        await CreateStageAndConnectAsync();

        // Then: 인증 전이므로 IsAuthenticated는 false여야 함
        Connector!.IsAuthenticated().Should().BeFalse("인증 전에는 IsAuthenticated가 false여야 함");
    }

    [Fact(DisplayName = "C-02-04: OnConnect 이벤트가 성공 결과로 발생한다")]
    public async Task OnConnect_Event_TriggersWithSuccess()
    {
        // Given: Stage 생성
        StageInfo = await TestServer.CreateTestStageAsync();

        var connectResult = false;
        var eventTriggered = false;
        var tcs = new TaskCompletionSource<bool>();

        Connector!.OnConnect += result =>
        {
            connectResult = result;
            eventTriggered = true;
            tcs.TrySetResult(result);
        };

        // When: 연결 시도
        Connector.Connect(
            TestServer.Host,
            TestServer.TcpPort,
            StageInfo.StageId,
            StageInfo.StageType
        );

        // OnConnect 이벤트 대기 (MainThreadAction 호출하면서 최대 5초)
        var completed = await WaitForConditionWithMainThreadActionAsync(() => tcs.Task.IsCompleted, 5000);

        // Then: 이벤트가 발생하고 성공 결과를 전달해야 함
        completed.Should().BeTrue("OnConnect 이벤트가 5초 이내에 발생해야 함");
        eventTriggered.Should().BeTrue("OnConnect 이벤트가 발생해야 함");
        connectResult.Should().BeTrue("연결 결과가 true여야 함");
    }

    [Fact(DisplayName = "C-02-05: 잘못된 Stage ID로 연결해도 TCP 연결은 성공한다")]
    public async Task Connect_WithInvalidStageId_TcpConnectionSucceeds()
    {
        // Given: 존재하지 않는 Stage ID
        var invalidStageId = 999999999L;

        // When: 잘못된 Stage ID로 연결 시도
        var connected = await Connector!.ConnectAsync(
            TestServer.Host,
            TestServer.TcpPort,
            invalidStageId,
            "TestStage"
        );

        // Then: TCP 연결 자체는 성공함 (Stage ID 검증은 서버 레벨에서 나중에 이루어짐)
        // PlayHouse 서버는 TCP 연결을 먼저 수락하고, 인증/Stage 참여 시 검증함
        connected.Should().BeTrue("TCP 연결 자체는 성공해야 함");
        Connector.IsConnected().Should().BeTrue("TCP 연결 상태여야 함");
        Connector.IsAuthenticated().Should().BeFalse("인증은 아직 안 됨");
    }

    [Fact(DisplayName = "C-02-06: 동일한 Connector로 재연결할 수 있다")]
    public async Task Connect_MultipleTimes_Succeeds()
    {
        // Given: 첫 번째 연결
        await CreateStageAndConnectAsync();
        Connector!.IsConnected().Should().BeTrue();

        // When: 연결 해제 후 재연결
        Connector.Disconnect();
        await Task.Delay(500); // 연결 해제 대기

        var newStageInfo = await TestServer.CreateTestStageAsync();
        var reconnected = await Connector.ConnectAsync(
            TestServer.Host,
            TestServer.TcpPort,
            newStageInfo.StageId,
            newStageInfo.StageType
        );

        // Then: 재연결이 성공해야 함
        reconnected.Should().BeTrue("재연결이 성공해야 함");
        Connector.IsConnected().Should().BeTrue("재연결 후 연결 상태가 true여야 함");
        Connector.StageId.Should().Be(newStageInfo.StageId, "새로운 Stage ID가 설정되어야 함");
    }
}
